crates/flox-rust-sdk/src/models/environment.rs

```diff
@@ -151,21 +152,40 @@ impl<S: TransactionState> PathEnvironment<S> {
     }
 
     /// Replace the contents of this environment's `.flox` with that of another environment's `.flox`
+    ///
+    /// This may copy build symlinks, so the assumption is that building self
+    /// will result in the same out link as building the replacement.
     pub fn replace_with(
         &mut self,
         replacement: PathEnvironment<Temporary>,
     ) -> Result<(), EnvironmentError2> {
-        fs_extra::dir::move_dir(
-            replacement.path,
-            &self.path,
-            &fs_extra::dir::CopyOptions::new()
-                .overwrite(true)
-                .content_only(true),
-        )
-        .expect("replace origin");
-        // std::fs::create_dir(&replacement.path).expect("recreate temp dir");
+        let transaction_backup = self
+            .path
+            .parent()
+            .expect("path shouldn't be /")
+            .join(format!("{}.tmp", self.name().0));
+        if transaction_backup.exists() {
+            return Err(EnvironmentError2::PriorTransaction(transaction_backup));
+        }
+        fs::rename(&self.path, &transaction_backup)
```

Comment:
Rename is a much more atomic operation than copy, this should either rename it or do nothing and error
We've been practicing `mv` swapping pretty much everywhere we have something transactional.

Reply:
I don't mean something wrong with this one `mv`, I mean something going wrong between steps of the sequence:
- Move `.flox` to a temporary directory, leaving the user _without_ a `.flox` for a moment
- Move the replacement's `.flox` into the original environment's `.flox`

If something goes wrong between those two steps the user is left without a `.flox`. Is it unlikely? Of course. But I consider this a bug waiting to be discovered and the fix is easy. I don't think we should ever leave the user without a `.flox` because then a bunch of our commands won't work at all.

Reply:
I did forget to try to move `name.tmp` back to `name` if moving tmp to `name` fails, so I can change that. But I believe it's pretty standard to implement transactions with move. If copy fails, we don't know if `.flox/name` contains the old contents, the new contents, or a little bit of both. If you're concerned about the edge case where it fails, we could add a check for `name.tmp` if we can't find `name` when opening an environment.

Reply:
I'm suggesting `copy` for starting the transaction and `move` for finishing it. I'm fine if we want to use `move` for both right now but I'll create a ticket to fix this later.

Reply:
Oh the temp is a **separate** temp, afaict

- We use copy to create the transaction state
- move the current state to a backup
- move the transaction state into the original place
- delete or reinstate the backup depending on the success of the former

Reply:
Pushed an attempt to put the `.tmp` dir back in place.

I think I understand what your saying is a bug, but as far as I'm aware, that's the most common approach to implementing transactions. We can discuss in a followup, but I did it that way deliberately.

Reply:
I consider this conversation resolved, let's move forward and address this later: https://github.com/flox/flox/issues/231